Mirror/Runtime/NetworkTransformBase.cs

```diff
@@ -0,0 +1,545 @@
+            Vector3 euler = rotation.eulerAngles;
+            if (compressRotation == Compression.None)
+            {
+                // write 3 floats = 12 byte
+                writer.Write(euler.x);
+                writer.Write(euler.y);
+                writer.Write(euler.z);
+            }
+            else if (compressRotation == Compression.Some)
+            {
+                // write 3 shorts = 6 byte. scaling [0,360] to [0,65535]
+                writer.Write(ScaleFloatToUShort(euler.x, 0, 360, ushort.MinValue, ushort.MaxValue));
+                writer.Write(ScaleFloatToUShort(euler.y, 0, 360, ushort.MinValue, ushort.MaxValue));
+                writer.Write(ScaleFloatToUShort(euler.z, 0, 360, ushort.MinValue, ushort.MaxValue));
+            }
+            else if (compressRotation == Compression.Much)
+            {
+                // write 3 byte. scaling [0,360] to [0,255]
+                writer.Write(ScaleFloatToByte(euler.x, 0, 360, byte.MinValue, byte.MaxValue));
+                writer.Write(ScaleFloatToByte(euler.y, 0, 360, byte.MinValue, byte.MaxValue));
+                writer.Write(ScaleFloatToByte(euler.z, 0, 360, byte.MinValue, byte.MaxValue));
+            }
+            else if (compressRotation == Compression.Lots)
+            {
+                // write 2 byte, 5 bits for each float
+                writer.Write(PackThreeFloatsIntoUShort(euler.x, euler.y, euler.z, 0, 360));
+            }
+
+            return true;
+        }
+
+        public override bool OnSerialize(NetworkWriter writer, bool initialState)
+        {
+            return SerializeIntoWriter(writer, initialState, targetComponent.transform.position, targetComponent.transform.rotation, compressRotation);
+        }
+
+        // try to estimate movement speed for a data point based on how far it
+        // moved since the previous one
+        // => if this is the first time ever then we use our best guess:
+        //    -> delta based on transform.position
+        //    -> elapsed based on send interval hoping that it roughly matches
+        static float EstimateMovementSpeed(DataPoint from, DataPoint to, Transform transform, float sendInterval)
+        {
+            Vector3 delta = to.position - (from != null ? from.position : transform.position);
+            float elapsed = from != null ? to.timeStamp - from.timeStamp : sendInterval;
+            return elapsed > 0 ? delta.magnitude / elapsed : 0; // avoid NaN
+        }
+
+        // serialization is needed by OnSerialize and by manual sending from authority
+        public void DeserializeFromReader(NetworkReader reader, bool initialState)
+        {
+            // put it into a data point immediately
+            DataPoint temp = new DataPoint();
+
+            // deserialize position
+            temp.position = reader.ReadVector3();
+
+            // deserialize rotation
+            if (compressRotation == Compression.None)
```

Comment:
refactor into:

Quaternion ReadRotation(NetworkReader reader, Compression compression)

Better yet,  add that to NetworkReader so it can be reused by any component that want to read/write a rotation